controller/orphan_controller.go

```diff
@@ -269,48 +303,205 @@ func (oc *OrphanController) isResponsibleFor(orphan *longhorn.Orphan) bool {
+}
+
+func (oc *OrphanController) extractOrphanedInstanceInfo(orphan *longhorn.Orphan) (name, instanceManager string, err error) {
+	name, ok := orphan.Spec.Parameters[longhorn.OrphanInstanceName]
+	if !ok {
+		return "", "", fmt.Errorf("failed to get instance name for instance orphan %v", orphan.Name)
+	}
+
+	instanceManager, ok = orphan.Spec.Parameters[longhorn.OrphanInstanceManager]
+	if !ok {
+		return "", "", fmt.Errorf("failed to get instance manager for instance orphan %v", orphan.Name)
+	}
+
+	switch orphan.Spec.DataEngine {
+	case longhorn.DataEngineTypeV1, longhorn.DataEngineTypeV2:
+		// supported data engine type
 	default:
-		err = fmt.Errorf("unknown orphan type %v", orphan.Spec.Type)
+		return "", "", fmt.Errorf("unknown data engine type %v for instance orphan %v", orphan.Spec.DataEngine, orphan.Name)
 	}
 
-	if err == nil || (err != nil && apierrors.IsNotFound(err)) {
-		return nil
+	return name, instanceManager, nil
+}
+
+func (oc *OrphanController) cleanupOrphanedInstance(orphan *longhorn.Orphan, instance, imName string, imType longhorn.InstanceManagerType, instanceCRSpec *longhorn.InstanceSpec) (isCleanupComplete bool, err error) {
+	if instanceCRSpec != nil && instanceCRSpec.NodeID == orphan.Spec.NodeID {
+		oc.logger.Infof("Orphan instance %v is scheduled back to current node %v. Skip cleaning up the instance resource and finalize the orphan CR.", instance, orphan.Spec.NodeID)
+		return true, nil
 	}
 
-	return err
+	imc, err := oc.getRunningInstanceManagerClientForOrphan(orphan, imName)
+	if err != nil {
+		return false, errors.Wrapf(err, "cannot get running instance manager client for orphan %v", orphan.Name)
+	} else if imc == nil {
+		oc.logger.WithField("orphanInstanceNode", orphan.Spec.NodeID).Warnf("No running instance manager for deleting orphan instance %v", orphan.Name)
+		return true, nil
+	}
+	defer func() {
+		if closeErr := imc.Close(); closeErr != nil {
+			oc.logger.WithError(closeErr).Error("Failed to close instance manager client")
+		}
+	}()
+
+	if err := oc.deleteInstance(imc, instance, imType, orphan.Spec.DataEngine); err != nil {
+		return false, err
+	}
+	isCleanupComplete, cleanupErr := oc.confirmOrphanInstanceCleanup(imc, instance, imType, orphan.Spec.DataEngine)
+	if cleanupErr == nil && !isCleanupComplete {
+		oc.logger.Infof("Orphan instance %v cleanup in progress", instance)
+	}
+	return isCleanupComplete, cleanupErr
 }
 
-func (oc *OrphanController) deleteOrphanedReplica(orphan *longhorn.Orphan) error {
+func (oc *OrphanController) getRunningInstanceManagerClientForOrphan(orphan *longhorn.Orphan, imName string) (*engineapi.InstanceManagerClient, error) {
+	im, err := oc.ds.GetRunningInstanceManagerByNodeRO(orphan.Spec.NodeID, orphan.Spec.DataEngine)
+	if err != nil {
+		oc.logger.WithError(err).Warnf("No running instance manager for node %v for deleting orphan instance %v", oc.controllerID, orphan.Name)
+		return nil, nil
```

Comment:
Do we need to handle different errors? For example, returns  `nil, nil`  only when the error is `not found`.